src/algorithms/AsymmetricForwardBackwardAdjoint.jl

```diff
@@ -0,0 +1,217 @@
+################################################################################
+# Primal-dual algorithms based on Asymmetric Forward-Backward-Adjoint that solve the following convex problem:
+#
+#								minimize_x f(x) + g(x) + (h □ l)(Lx), 
+#
+# where f is smooth, g and h are possibly nonsmooth and l is strongly convex. The infimal convolution is denoted by '□', and L is a linear mapping.   
+#
+# Usage:	 it, x, sol = ProximalAlgorithms.AFBA(x0, y0; g, h, f, l, betaQ, betaR, gamma1, gamma2, theta, mu)   
+#
+# The iterator implements Algorithm 3 of [1] with constant stepsize (α_n=λ) for several prominant special cases:
+# 1) θ = 2 	 	==>   Corresponds to the Vu-Condat Algorithm [2,3].
+# 2) θ = 1, μ=1
+# 3) θ = 0, μ=1
+# 4) θ ∈ [0,∞), μ=0 
+#
+# See [1, Figure 1] for other special cases and relation to other algorithms.   
+#
+# [1] Latafat, Patrinos. "Asymmetric forward–backward–adjoint splitting for solving monotone inclusions involving three operators"  Computational Optimization and Applications, pages 1–37, 2017.
+# [2] Condat. "A primal–dual splitting method for convex optimization involving Lipschitzian, proximable and linear composite terms" Journal of Optimization Theory and Applications 158.2 (2013): 460-479.
+# [3] Vũ. "A splitting algorithm for dual monotone inclusions involving cocoercive operators"" Advances in Computational Mathematics, 38(3), pp.667-681.
+
+struct AFBAIterator{I <: Integer, R <: Real, T1 <: BlockArray, T2 <: BlockArray} <: ProximalAlgorithm{I, T1}
+    x::T1
+    y::T2
+    g
+    h
+    f
+    l
+    L
+    theta::R
+    mu::R
+    lam::R
+    betaQ::R # Lipschitz constant of f
+    betaR::R # Lipschitz constant of l
+    gamma1::R
+    gamma2::R
+    maxit::I
+    tol::R
+    verbose::I
+    verbose_freq::I
+    xbar::T1
+    ybar::T2
+    gradf::T1
+    gradl::T2
+    FPR_x::T1
+    FPR_y::T2
+    temp_x::T1
+    temp_y::T2
+end
+
+################################################################################
+# Constructor(s)
+
+function AFBAIterator(x0::T1, y0::T2; g=Zero(), h=Zero(), f=Zero(), l=Zero(), L=Identity(blocksize(x0)), theta=1, mu=1, lam=1, betaQ=0, betaR=0, gamma1=-1, gamma2=-1, maxit::I=10000, tol::R=1e-5, verbose=1, verbose_freq = 100) where {I, R, T1, T2}
```

Comment:
Last modifications after which we can merge:
* Have you tried using the following defaults here?
```
g=IndFree(), h=IndFree(), f=IndFree(), l=IndZero()
```
`IndFree` is the identically zero function from `ProximalOperators`. If you do that, you should be able to eliminate the special case of `h` being zero (since `IndFree` you can `Conjugate` without problem). Also, this way argument `l` is really ell, not ell-conjugate, which is how it should be.
* You can construct `Conjugate(h)` and `Conjugate(l)` here once and store them in the iterator, to use them later in `iterate`.

Reply:
I see. What you can do is to define what the `Conjugate` constructor does for these two special functions, that is 
```
Conjugate(f::IndFree) = IndZero()
Conjugate(f::IndZero) = IndFree()
```
That should be sufficient, I guess? It also makes these "trivial" conjugations direct, without applying Moreau identity to compute the prox in these simple cases.

You can put these definitions in a new file src/utilities/conjugate.jl (and remember to include this from ProximalAlgorithms.jl).

Reply:
Thanks for pointing this out @pylat, this was very helpful feedback. I updated the `master` branch with these conjugation shortcuts and now everything should be much easier: If you pull it and `git --rebase` your branch on top of it, you'll see that you can `Conjugate` the `IndZero` and `IndFree` and compute `gradient!` and `prox!` with no problem, like in [this test](https://github.com/kul-forbes/ProximalAlgorithms.jl/blob/3a38883af55243109fed2e2d0eab47432d6052b9/test/test_conjugate.jl).

Reply:
great! So now I we should also add some of the famous conjugates in the conjugate.jl file so as to be able to evaluate the gradient easily for conjugate of l

Reply:
One last issue is that when h\equiv 0, L=0 not identity. We should fix this to be able to remove the if in the iterator. 

 We need a zero linear operator like identity.jl. Since we are using IndFree() in place of  Zero() for the functions should I modify zero.jl to define A_mul_B! and such for it?

Reply:
I don't get it: why do you need to separate the case h \equiv 0 and set L = 0 in that case? If h \equiv 0 then L can be whatever: the identity for example, which is also a desirable default value when the user does not specify L.

Reply:
Also: let's first have the algorithm there nicely, close this PR, and then go on with other stuff (like conjugations) separately.

Reply:
The issue is the last step. If h = 0 and L=id then we have ybar^k = 0, and 
y^{k+1} = bary^k + gamma2*(1-\mu)*(2-\theta) L(\barx^k - x^k) which is not zero. Am I missing something?

Reply:
It will be zero iff the primal residual is zero. Isn't that ok?

Reply:
you are right. I was assuming that when h=0 we must get FBS. But yes actually we just get some strange iteration. Unless theta=2 or mu=1 in which case we get FBS. 

The thing is that if h=IndFree and l=IndZero then does it even make sense to use AFBA and performing this useless dual update. Shall we raise an error in that case and suggest using FBS?

Reply:
I updated the files. I guess what is left is to add useful Conjugation rules so as to be able to compute gradient of conjugate of l. But this can be done later.

Reply:
Yeah I'm against raising errors with questionable suggestions to the user: the user invokes an algorithm with certain parameters, he gets what the algorithm does with those parameters.